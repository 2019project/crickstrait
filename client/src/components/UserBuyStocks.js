import React, { Component } from "react";
import UserNavigation from "./Common/UserNavigation";
import "../styles/UserBuyStocks.css";
import { connect } from "react-redux";
import { getAllStocks, sellStocks } from "../actions/Users";
import jwt_decode from "jwt-decode";

export class UserBuyStocks extends Component {
  state = {
    qty: "",
    total: 0
  };

  componentDidMount() {
    if (localStorage.getItem("token")) {
      var decode = jwt_decode(localStorage.getItem("token"));
      let email = {
        email: decode.email
      };
      this.props.getAllStocks(email);
    }
  }

  render() {
    if (localStorage.getItem("token")) {
      var decode = jwt_decode(localStorage.getItem("token"));
    }
    var rand;
    return (
      <div>
        <div id="buyStocksContainer">
          <UserNavigation selected="buy" />
          <div id="userStocksContainer">
            <table id="userBuyStocksTable">
              <thead id="userBuyStocksHeader">
                <th>Stocks</th>
                <th>Bought Price</th>
                <th>Quantity</th>
                <th>Selling price</th>
                <th></th>
                {/* <th>Date</th> */}
                <th></th>
              </thead>
              {this.props.users.map(hist => (
                <>
                  {hist.company.map((stocks, index) => (
                    <tr>
                      <td id={"tickerName" + index}>{stocks.ticker_name}</td>
                      <td id={"tickerPrice" + index}>{stocks.current_price}</td>

                      <td id={"tickerQty" + index}>{stocks.buying_quantity}</td>
                      <td>
                        {" "}
                        {""}
                        {
                          (rand = Number(
                            Math.random() *
                              (Number(stocks.current_price) +
                                5 -
                                (Number(stocks.current_price) - 5)) +
                              (Number(stocks.current_price) - 5)
                          ).toFixed(2))
                        }
                      </td>
                      <td>
                        {stocks.current_price > rand ? (
                          <span
                            class="fa fa-caret-down"
                            style={{ color: "red", fontSize: "18px" }}
                          ></span>
                        ) : (
                          <span
                            class="fa fa-caret-up"
                            style={{ color: "green", fontSize: "18px" }}
                          ></span>
                        )}{" "}
                        {Number(
                          ((rand - stocks.current_price) /
                            stocks.current_price) *
                            100
                        ).toFixed(2)}{" "}
                        %
                      </td>
                      <td>
                        <div class="sellStocksBox">
                          <a class="sell" href="#sellStockspopup1">
                            Sell
                          </a>
                        </div>

                        <div id="sellStockspopup1" class="sellStocksoverlay">
                          <div class="sellStockspopup">
                            <h2>{stocks.ticker_name} </h2>
                            <a class="sellStocksclose" href="#">
                              &times;
                            </a>
                            <div class="sellStockscontent">
                              <p>Current Price : ${rand}</p>
                              <p>
                                Quantity :{" "}
                                <input
                                  type="text"
                                  id="sellingQuantity"
                                  name="qty"
                                  value={this.state.qty}
                                  placeholder="0"
                                  onChange={e =>
                                    this.setState({
                                      [e.target.name]: e.target.value,
                                      total: (e.target.value *= Number(213.55))
                                    })
                                  }
                                />
                              </p>
                              <p>Total Price : ${this.state.total}</p>
                              <a
                                id="sellSpecificStock"
                                href="#"
                                onClick={() => {
                                  {
                                    let user = {
                                      email: decode.email,
                                      ticker_name: stocks.ticker_name,
                                      sell_price: rand,
                                      qty: this.state.qty,
                                      price: this.state.total
                                    };
                                    this.props.sellStocks(user);
                                  }
                                }}
                              >
                                Sell
                              </a>
                            </div>
                          </div>
                        </div>
                      </td>
                    </tr>
                  ))}
                </>
              ))}
            </table>
            {/* ) : (
                <h1>No Stocks History</h1>
              )} */}
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  users: state.usersReducer.users,
  isLoading: state.LoadingReducer.isLoading
});
export default connect(mapStateToProps, {
  getAllStocks,
  sellStocks
})(UserBuyStocks);
