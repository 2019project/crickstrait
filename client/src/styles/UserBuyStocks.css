@import url(https://fonts.googleapis.com/css?family=Karla);
* {
  font-family: Karla;
}

#buyStocksContainer {
  width: 100%;
  display: grid;
  grid-template-rows: 150px auto;
  text-align: center;
}

#userStocksContainer {
  background: url("../components/Common/stocks_wallpaper.jpg") no-repeat;
  background-size: cover;
  margin-top: 60px;
  height: 92vh;
}

#userStocksContainer h1 {
  text-align: center;
  font-weight: 700;
  font-family: Karla;
  font-size: 40px;
  margin-top: 200px;
  color: #2c3e50;
  text-shadow: 3px 3px 6px #707070;
}

.sellStocksBox {
  margin: 0 auto;
  border: 2px solid #fff;
  text-align: center;
}

.sell {
  font-size: 1em;
  border-radius: 20px;
  padding: 5px 20px;
  color: #fff;
  background-color: #39abf7;
  text-decoration: none;
  cursor: pointer;
  transition: all 0.3s ease-out;
}
.sell:hover {
  background: #2c3e50;
}

.sellStocksoverlay {
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background: rgba(0, 0, 0, 0.7);
  transition: opacity 500ms;
  visibility: hidden;
  opacity: 0;
}
.sellStocksoverlay:target {
  visibility: visible;
  opacity: 1;
}

.sellStockspopup {
  margin: 150px auto;
  padding: 8px;
  background: #2c3e50;
  border-radius: 5px;
  width: 95%;
  position: relative;
  transition: all 5s ease-in-out;
}

.sellStockspopup h2 {
  margin-top: 0;
  color: #fff;
  font-family: Tahoma, Arial, sans-serif;
  font-weight: 500;
  font-size: 40px;
}
.sellStockspopup .sellStocksclose {
  position: absolute;
  top: 20px;
  right: 30px;
  transition: all 200ms;
  font-size: 30px;

  font-weight: bold;
  text-decoration: none;
  color: #fff;
}
.sellStockspopup .sellStocksclose:hover {
  color: #39abf7;
  cursor: pointer;
}
.sellStockspopup .sellStockscontent {
  max-height: 30%;
  /* overflow: auto; */
  color: #fff;
}

.sellStockspopup .sellStockscontent p {
  font-size: 18px;
  margin: 25px auto;
}

#userBuyStocksTable {
  width: 90%;
  height: auto;
  background-color: #fff;
  border-collapse: collapse;
  border-radius: 4px;
  font-size: 8px;
  box-shadow: 4px 4px 6px #707070;
  margin: 30px auto;
}
#userBuyStocksHeader {
  background-color: #2c3e50;
  color: #fff;
  border-radius: 8px;
  width: 70%;
}

#userBuyStocksTable th {
  padding: 15px;
  color: #fff;
}

#userBuyStocksTable td {
  padding: 10px;
  text-align: center;
  color: #2c3e50;
  font-weight: 550;
}

#sellSpecificStock {
  border: 2px solid #39abf7;
  padding: 5px 25px;
  color: #fff;
  outline: none;
  margin: 10px 0;
  font-family: Karla;
  text-decoration: none;
  cursor: pointer;
  border-radius: 4px;
  background: transparent;
}

#sellSpecificStock:hover {
  right: 0;
  background-color: #39abf7;
  box-shadow: 0 12px 16px 0 rgba(231, 226, 226, 0.24),
    0 17px 50px 0 rgba(231, 226, 226, 0.24);
}

#sellingQuantity {
  width: 18%;
  height: 20px;
  text-align: center;
  line-height: 1.65;
  background: transparent;
  padding: 13px;
  font-weight: lighter;
  color: #fff;
  border: 2px solid #39abf7;
  border-radius: 4px;
}

#sellingQuantity::placeholder {
  color: #7a7a7a;
  font-weight: lighter;
}

@media screen and (min-width: 430px) {
  #buyStocksContainer {
    width: 100%;
    grid-template-columns: 250px 1fr;
    font-family: Karla;
    margin: auto;
  }

  #userStocksContainer {
    margin-top: 0px;
  }

  .sellStockspopup {
    margin: 250px auto;
    padding: 20px;
    background: #2c3e50;
    border-radius: 5px;
    width: 28%;
    position: relative;
    transition: all 5s ease-in-out;
  }

  #userBuyStocksTable {
    width: 50%;
    font-size: 16px;
    height: auto;
    background-color: #fff;
    border-collapse: collapse;
    border-radius: 4px;
    margin: 100px auto;
  }
}
