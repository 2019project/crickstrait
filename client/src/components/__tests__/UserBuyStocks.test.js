import React from "react";
import { shallow, mount } from "enzyme";
import { UserBuyStocks } from "../UserBuyStocks";

const userBuyStocks = jest.fn();
const sellStocks = jest.fn();
const getAllStocks = jest.fn();
let total = 0;
const users = [
  {
    company: [
      {
        buy: true,
        buy_date: "2019-11-22T07:27:35.875Z",
        buying_quantity: "7",
        current_price: 201.55,
        sell: false,
        ticker_name: "AAPL"
      }
    ]
  }
];

const wrapper = shallow(
  <UserBuyStocks
    userBuyStocks={userBuyStocks}
    users={users}
    sellStocks={sellStocks}
    getAllStocks={getAllStocks}
  />
);

describe("Testing UserbuyStocks Component", () => {
  it("should mount the component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should have a tickername in each row ", () => {
    console.log(localStorage.getItem("token"));
    expect(wrapper.find("#tickerName0").props().children).toBe(
      users[0].company[0].ticker_name
    );
  });

  it("should have a tickerPrice in each row ", () => {
    expect(wrapper.find("#tickerPrice0").props().children).toBe(
      users[0].company[0].current_price
    );
  });

  it("should have a tickerOty in each row ", () => {
    expect(wrapper.find("#tickerQty0").props().children).toBe(
      users[0].company[0].buying_quantity
    );
  });

  it("shoudl check for input value", () => {
    const e = {
      target: {
        name: "total",
        value: 0
      }
    };
    expect(wrapper.state().total).toBe(e.target.value);
  });

  it("should simulate on Chnage on input change ", () => {
    wrapper
      .find("#sellingQuantity")
      .simulate("change", { target: { value: 0 } });
    expect(wrapper.state().total).toBe(0);
  });

  it("should simulate onClick on sell stock button ", () => {
    wrapper.find("#sellSpecificStock").simulate("click");
  });

  it("should get toekn from localstorage ", () => {
    expect(localStorage.getItem("token")).toBe(null);
  });
});
