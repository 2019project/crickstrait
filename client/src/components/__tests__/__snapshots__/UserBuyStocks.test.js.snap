// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing UserbuyStocks Component should mount the component 1`] = `
<div>
  <div
    id="buyStocksContainer"
  >
    <Connect(UserNavigation) />
    <div
      id="userStocksContainer"
    >
      <table
        id="userBuyStocksTable"
      >
        <thead
          id="userBuyStocksHeader"
        >
          <th>
            Stocks
          </th>
          <th>
            Bought Price
          </th>
          <th>
            Quantity
          </th>
          <th>
            Selling price
          </th>
          <th />
          <th />
        </thead>
        <tr>
          <td
            id="tickerName0"
          >
            AAPL
          </td>
          <td
            id="tickerPrice0"
          >
            201.55
          </td>
          <td
            id="tickerQty0"
          >
            7
          </td>
          <td>
             
            197.41
          </td>
          <td>
            <span
              class="fa fa-caret-down"
              style={
                Object {
                  "color": "red",
                  "fontSize": "18px",
                }
              }
            />
             
            -2.05
             
            %
          </td>
          <td>
            <div
              class="sellStocksBox"
            >
              <a
                class="sell"
                href="#sellStockspopup1"
              >
                Sell
              </a>
            </div>
            <div
              class="sellStocksoverlay"
              id="sellStockspopup1"
            >
              <div
                class="sellStockspopup"
              >
                <h2>
                  AAPL
                   
                </h2>
                <a
                  class="sellStocksclose"
                  href="#"
                >
                  ×
                </a>
                <div
                  class="sellStockscontent"
                >
                  <p>
                    Current Price : $
                  </p>
                  <p>
                    Quantity :
                     
                    <input
                      id="sellingQuantity"
                      name="qty"
                      onChange={[Function]}
                      placeholder="0"
                      type="text"
                      value=""
                    />
                  </p>
                  <p>
                    Total Price : $
                  </p>
                  <a
                    href="#"
                    id="sellSpecificStock"
                  >
                    Buy
                  </a>
                </div>
              </div>
            </div>
          </td>
        </tr>
        <tr>
          <td
            id="tickerName1"
          >
            MSFT
          </td>
          <td
            id="tickerPrice1"
          >
            135.68
          </td>
          <td
            id="tickerQty1"
          >
            6
          </td>
          <td>
             
            138.66
          </td>
          <td>
            <span
              class="fa fa-caret-up"
              style={
                Object {
                  "color": "green",
                  "fontSize": "18px",
                }
              }
            />
             
            2.20
             
            %
          </td>
          <td>
            <div
              class="sellStocksBox"
            >
              <a
                class="sell"
                href="#sellStockspopup1"
              >
                Sell
              </a>
            </div>
            <div
              class="sellStocksoverlay"
              id="sellStockspopup1"
            >
              <div
                class="sellStockspopup"
              >
                <h2>
                  MSFT
                   
                </h2>
                <a
                  class="sellStocksclose"
                  href="#"
                >
                  ×
                </a>
                <div
                  class="sellStockscontent"
                >
                  <p>
                    Current Price : $
                  </p>
                  <p>
                    Quantity :
                     
                    <input
                      id="sellingQuantity"
                      name="qty"
                      onChange={[Function]}
                      placeholder="0"
                      type="text"
                      value=""
                    />
                  </p>
                  <p>
                    Total Price : $
                  </p>
                  <a
                    href="#"
                    id="sellSpecificStock"
                  >
                    Buy
                  </a>
                </div>
              </div>
            </div>
          </td>
        </tr>
      </table>
    </div>
  </div>
</div>
`;
